text/0091-ci-upload-tokens.md

```diff
@@ -0,0 +1,271 @@
+
+# Technical Implementation
+
+The motivation is to add a new kind of token to Sentry which are fundamentally per-organization
+tokens, but with the ability to carry meta information that tools like sentry-cli can use to
+improve the user experience.  These org level tokens can be created by anyone in the org, they
+can be given additional restrictions, and they can carry meta information such as routing
+data.  For the purpose of this document they are called **structural tokens**.
+
+## Token Format
+
+The proposed token format is to leverage JWT as serialization format.  The goals of the
+token align generally with both [Macaroons](http://macaroons.io/) and
+[Biscuit](https://www.biscuitsec.org) but unfortunately the former standard has never seen
+much attention, and the latter is pretty new, not particularly proven and very complex.
+Either system however permits adding additional restrictions to the token which make them
+a very interesting choice for the use in our pipeline.
+
+One of the benefits of having the tokens carry additional data is that the token alone has enough
+information available to route to a Sentry installation.  This means that `sentry-cli` or
+any other tool _just_ needs the token to even determine the host that the token should be
+sent against.  This benefit also applies to JWT or PASETO tokens which can be considered
+for this as well.  The RFC here thus proposes to encode this data into a regular **JWT**
+token.
+
+A serialized token is added a custom prefix `sntrys_` (sentry structure) to make
+it possible to detect it by security scrapers.  Anyone handling such a token is
+required to check for the `sntrys_` prefix and disregard it before parsing it.  This
+can also be used by the client side to detect a structural token if the client is
+interested in extracting data from the token.
+
+## Token Facts
+
+We want to encode certain information into the tokens.  The following attributes are defined:
+
+* `sentry_site`: references the target API URL that should be used.  A token will always have a
+  site in it and clients are not supposed to provide a fallback.  For instance this
+  would be `https://myorg.sentry.io/`.
+* `sentry_org`: a token is uniquely bound to an org, so the slug of that org is also always
+  contained.  Note that the slug is used rather than an org ID as the clients typically
+  need these slugs to create API requests.
+
+Potential fields:
+
+* `sentry_projects`: normally a token is valid for the entire org, but it could potentially
+  be restricted.  For operations such as source map uploads it might be interesting to issue
+  tokens bound to a single project in which case the upload experience does not require
+  providing the project slugs.  However we currently do not want to start with this.
+
+For JWT the facts could be encoded as custom claims:
+
+```json
+{
+    "iss": "sentry.io",
+    "iat": 1684154626,
+    "sentry_site": "https://myorg.sentry.io/",
+    "sentry_org": "myorg"
+}
```

Comment:
We should add a `nonce` field here so these are less predictable/guessable. With this current format, someone would just need to brute force the `iat` to find a value token.

---

We should also store the tokens as hashed values in the database.

Then when it comes to verifying an inbound request, it's a matter of:
1. hash the provided token
2. check the database for a hash of equal value
3. if one is found, decode the provided token
5. perform requested operation (pending authorization)

Reply:
Are the last characters to help identify the token? If so, Could we instead expand the model to have a `name` or `description` or both? Then the user can provide a meaningful value to them.

Reply:
Yeah, that is okay. 👍